UMLEditor/UMLEditor/Classes/TimeMachine.cs

```diff
@@ -7,93 +8,58 @@ namespace UMLEditor.Classes;
 /// </summary>
 public static class TimeMachine
 {
-    /// <summary>
-    /// Node class to represent a single state in the time machine history.
-    /// </summary>
-    public class Node
-    {
-        /// <summary>
-        /// Reference to previous node
-        /// </summary>
-        public Node? PrevNode;
-        /// <summary>
-        /// Diagram at this node
-        /// </summary>
-        public Diagram? StateDiagram;
-        /// <summary>
-        /// Reference to next node
-        /// </summary>
-        public Node? NextNode;
-    }
-    /// <summary>
-    /// Reference to the beginning state of the time machine
-    /// </summary>
-    public static Node? Head { get; private set; }
-    /// <summary>
-    /// Reference to the end state of the time machine
-    /// </summary>
-    public static Node? Tail { get; private set; }
-    /// <summary>
-    /// Reference to the current state of the time machine
-    /// </summary>
-    public static Node? Current { get; private set; }
-    /// <summary>
-    /// Counter for number of states in the time machine
-    /// </summary>
-    public static int Size { get; private set; }
+    private static List<Diagram> _diagramList = new List<Diagram>();
+
+    private static int _head = -1, _current = -1, _tail = -1;
 
     /// <summary>
     /// Adds a new diagram to the end of the list. Erases redo history if diagram is added to middle of list.
     /// </summary>
     /// <param name="d">Diagram to add</param>
     public static void AddState(Diagram d)
     {
-        // Add new node to end of list
-        Node newNode = new Node();
-        newNode.StateDiagram = new Diagram(d);
-        newNode.PrevNode = Current;
-        newNode.NextNode = null;
-        if (Current is not null)
+        if (_current != _tail)
         {
-            Current.NextNode = newNode;
+            _diagramList.RemoveRange(_current + 1, _tail);
```

Comment:
Small bug. The second parameter to RemoveRange is the number of elements to remove. _tail is going to go past the end of the array.